src/indexer/demuxer.rs

```diff
@@ -0,0 +1,343 @@
+    /// the via `set` method.
+    pub fn new(num_docids: usize) -> Self {
+        let mut vec = vec![];
+        vec.resize(num_docids, 0);
+
+        DocidToSegmentOrdinal {
+            docid_index_to_segment_ordinal: vec,
+        }
+    }
+
+    /// Associated a docids with a the new SegmentOrdinal.
+    pub fn set(&mut self, docid: u32, segment_ordinal: SegmentOrdinal) {
+        self.docid_index_to_segment_ordinal[docid as usize] = segment_ordinal;
+    }
+
+    /// Iterates over the new SegmentOrdinal in the order of the docid.
+    pub fn iter(&self) -> impl Iterator<Item = &SegmentOrdinal> {
+        self.docid_index_to_segment_ordinal.iter()
+    }
+}
+
+impl DemuxMapping {
+    /// Creates a new empty DemuxMapping.
+    pub fn empty() -> Self {
+        DemuxMapping {
+            mapping: Default::default(),
+        }
+    }
+
+    /// Creates a DemuxMapping from existing mapping data.
+    pub fn new(mapping: Vec<DocidToSegmentOrdinal>) -> Self {
+        DemuxMapping { mapping }
+    }
+
+    /// Adds a DocidToSegmentOrdinal. The order of the pus calls
+    /// defines the old segment ordinal. e.g. first push = ordinal 0.
+    pub fn add(&mut self, segment_mapping: DocidToSegmentOrdinal) {
+        self.mapping.push(segment_mapping);
+    }
+
+    /// Returns the old number of segments.
+    pub fn get_old_num_segments(&self) -> usize {
+        self.mapping.len()
+    }
+}
+
+fn get_delete_bitsets(
+    demux_mapping: &DemuxMapping,
+    target_segment_ordinal: SegmentOrdinal,
+    max_value_per_segment: &[u32],
+) -> Vec<AliveBitSet> {
+    let mut bitsets: Vec<_> = max_value_per_segment
+        .iter()
+        .map(|max_value| BitSet::with_max_value_and_full(*max_value))
+        .collect();
+
+    for (old_segment_ordinal, docid_to_new_segment) in demux_mapping.mapping.iter().enumerate() {
+        let bitset_for_segment = &mut bitsets[old_segment_ordinal];
+        for docid in docid_to_new_segment
+            .iter()
+            .enumerate()
+            .filter(|(_docid, new_segment_ordinal)| **new_segment_ordinal != target_segment_ordinal)
+            .map(|(docid, _)| docid)
+        {
+            // mark document as deleted if segment ordinal is not target segment ordinal
+            bitset_for_segment.remove(docid as u32);
+        }
+    }
+
+    bitsets
+        .iter()
+        .map(|bitset| AliveBitSet::from_bitset(bitset))
+        .collect_vec()
+}
+
+/// Demux the segments according to `demux_mapping`. See `DemuxMapping`.
+/// The number of output_directories need to match max new segment ordinal from `demux_mapping`.
+///
+/// The ordinal of `segments` need to match the ordinals provided in `demux_mapping`.
+pub fn demux<Dir: Directory>(
+    segments: &[Segment],
+    demux_mapping: &DemuxMapping,
+    target_settings: IndexSettings,
+    mut output_directories: Vec<Dir>,
+) -> crate::Result<Vec<Index>> {
+    output_directories.reverse();
```

Comment:
I don't think you need to inverse the Vec here.
```
let output_directories_it = output_directories.into_iter()
```


```
output_directories_it.next().ok_or_else(|| ...)
```

Reply:
Or even better, I believe you could do 


```
for (target_segment_ordinal, output_directory) in output_directories.into_iter().enumerate() {
   // ...
}
```

Reply:
yep that's much better